python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -63,26 +63,58 @@
 def broadcast(
     *columns: NamedColumn, target_length: int | None = None
 ) -> list[NamedColumn]:
-    lengths = {column.obj.size() for column in columns}
-    if len(lengths - {1}) > 1:
-        raise RuntimeError("Mismatching column lengths")
+    """
+    Broadcast a sequence of columns to a common length.
+
+    Parameters
+    ----------
+    columns
+        Columns to broadcast.
+    target_length
+        Optional length to broadcast to. If not provided, uses the
+        non-unit length of existing columns.
+
+    Returns
+    -------
+    List of broadcasted columns all of the same length.
+
+    Raises
+    ------
+    RuntimeError
+        If broadcasting is not possible.
+
+    Notes
+    -----
+    In evaluation of a set of expressions, polars type-puns length-1
+    columns with scalars. When we insert these into a DataFrame
+    object, we need to ensure they are of equal length. This function
+    takes some columns, some of which may be length-1 and ensures that
+    all length-1 columns are broadcast to the length of the others.
+
+    Broadcasting is only possible if the set of lengths of the input
+    columns is a subset of ``{1, n}`` for some (fixed) ``n``. If
+    ``target_length`` is provided and not all columns are length-1
+    (i.e. ``n != 1``), then ``target_length`` must be equal to ``n``.
+    """
+    lengths: Set[int] = {column.obj.size() for column in columns}
     if lengths == {1}:
         if target_length is None:
             return list(columns)
         nrows = target_length
-    elif len(lengths) == 1:
-        if target_length is not None:
-            assert target_length in lengths
-        return list(columns)
     else:
-        (nrows,) = lengths - {1}
-        if target_length is not None:
-            assert target_length == nrows
+        try:
+            (nrows,) = lengths - {1}
```

Comment:
Could be explicit and call `.difference`, not sure if everyone knows how `-` is overridden for sets.

Reply:
Done, needed to type with `set`, not `AbstractSet` because that doesn't advertise `difference` as a method :exploding_head: